istioctl/pkg/auth/upgrader.go

```diff
@@ -0,0 +1,376 @@
+	}
+	if v1alpha1Model.Permissions == nil || len(v1alpha1Model.Permissions) == 0 {
+		return fmt.Errorf("permissions are empty")
+	}
+	if v1alpha1Model.Principals == nil || len(v1alpha1Model.Principals) == 0 {
+		return fmt.Errorf("principals are empty")
+	}
+	// TODO(pitlv2109): Support more complex cases
+	if len(v1alpha1Model.Permissions) > 1 {
+		return fmt.Errorf("more than one access rule is not supported")
+	}
+	accessRule := v1alpha1Model.Permissions[0]
+	operations := convertAccessRuleToOperation(&accessRule)
+	if operations == nil {
+		return fmt.Errorf("cannot convert access rule to operation")
+	}
+	sources := convertBindingToSources(v1alpha1Model.Principals)
+	// If there is no services field or it's defined as "*", it's equivalent to an AuthorizationPolicy
+	// with no workload selector
+	authzPolicyConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      schemas.AuthorizationPolicy.Type,
+			Namespace: namespace,
+		},
+		Spec: &rbac_v1beta1.AuthorizationPolicy{
+			Selector: &v1beta1.WorkloadSelector{},
+			Rules: []*rbac_v1beta1.Rule{
+				{
+					From: sources,
+					To: []*rbac_v1beta1.Rule_To{
+						{
+							Operation: operations,
+						},
+					},
+				},
+			},
+		},
+	}
+	if len(accessRule.Services) == 0 || accessRule.Services[0] == "*" {
+		authzPolicyConfig.Name = fmt.Sprintf("%s-services", namespace)
+		authzPolicyConfig.Spec.(*rbac_v1beta1.AuthorizationPolicy).Selector = nil
+		ug.AuthorizationPolicies = append(ug.AuthorizationPolicies, authzPolicyConfig)
+		return nil
+	}
+	for _, service := range accessRule.Services {
+		serviceName := strings.Split(service, ".")[0]
+		authzPolicyConfig.Name = fmt.Sprintf("%s-%s", serviceName, namespace)
+		workloadSelector, err := ug.getAndAddServiceToWorkloadLabelMapping(service, namespace)
+		if err != nil {
+			return err
+		}
+		authzPolicyCopy, err := copystructure.Copy(&authzPolicyConfig)
+		if err != nil {
+			return err
+		}
+		authzPolicyCopy.(*model.Config).Spec.(*rbac_v1beta1.AuthorizationPolicy).Selector.MatchLabels = workloadSelector
+		authzPolicyConfigCopy := authzPolicyCopy.(*model.Config)
+		ug.AuthorizationPolicies = append(ug.AuthorizationPolicies, *authzPolicyConfigCopy)
+	}
+	return nil
+}
+
+// TODO(pitlv2109): Handle cases with workload selector from destination.labels and other constraints
+// convertAccessRuleToOperation converts one Access Rule to the equivalent Operation.
+func convertAccessRuleToOperation(rule *authz_model.Permission) *rbac_v1beta1.Operation {
+	if rule == nil {
+		return nil
+	}
+	operation := rbac_v1beta1.Operation{}
+	operation.Methods = rule.Methods
+	operation.Paths = rule.Paths
+	// TODO(pitlv2109): Handle destination.port
+	return &operation
+}
+
+// TODO(pitlv2109): Handle properties that are not promoted to first class fields.
+// convertBindingToSources converts Subjects to the equivalent Sources.
+func convertBindingToSources(principals []authz_model.Principal) []*rbac_v1beta1.Rule_From {
```

Comment:
It seems that we only supports "properties" but not "constraints"?

Reply:
Yes this only converts fields from `properties` that are promoted to first class fields.